libraries/cyclestreets-view/src/main/java/net/cyclestreets/content/DatabaseHelper.java

```diff
@@ -46,6 +48,16 @@
 
   DatabaseHelper(final Context context) {
     super(context, DATABASE_NAME, null, DATABASE_VERSION);
+
+    // From Android 9 (API version 28, SQLite version 3.22.0) Compatibility WAL is enabled by default.
+    // This breaks our testing mechanism of pasting over a new version of the database, so disable
+    // it when we're in debug mode.  See:
+    // -  https://source.android.com/devices/tech/perf/compatibility-wal
+    // -  https://www.sqlite.org/wal.html
+    // -  https://www.sqlite.org/tempfiles.html
+    if (BuildConfig.DEBUG) {
+      this.setWriteAheadLoggingEnabled(false);
+    }
```

Comment:
Is this only required for the tests?  If so, is it possible to move it into the test fixture?

Reply:
I tried that.  Unfortunately, what that means is that when the app starts up (before this point) and initialises the DB, somehow SQLite tracks the state that it's running in WAL mode. So when you then replace the DB (even if WAL mode is disabled in the tests, and regardless of whether you leave or delete the `-shm` and `-wal` files), then SQLite fails to get a readable or writeable database, claiming it's already locked. 😞 

So unfortunately, in order to keep the ability to test in this way (which was incredibly useful when making sure DB migration didn't break anything) we do have to have this tweak to our code.  However, since it's wrapped in the if test, it'll be compiled out in release mode, so it's just pollution to our source code and not the final artifact.

Reply:
I did wonder if there was a sequencing issue.